### PR TITLE
Rewrite tests and example for CustomSet

### LINKED ISSUE #121: Change use of Set and HashSet to MapSet per the Elixir 1.2 changelog
Per https://github.com/elixir-lang/elixir/blob/v1.2.0/CHANGELOG.md

```
allergies/allergies_test.exs
77:    set = Enum.into(list, HashSet.new)
79:      |> Enum.into(HashSet.new)
80:      |> Set.equal?(set)
82:           "Expected a set with: #{inspect to_contain} got #{inspect set |> Set.to_list}"

custom-set/custom_set.exs
1:defmodule CustomSet do
2:  # This lets the compiler check that all Set callback functions have been
4:  @behaviour Set

custom-set/custom_set_test.exs
10:defmodule CustomSetTest do
13:  # Tip: to make your set work with the Set module be sure that your set value
14:  # is a tuple with as first element `CustomSet`. A record with tag `CustomSet`
19:    assert Set.equal?(Set.delete(CustomSet.new([3,2,1]), 2), CustomSet.new([1,3]))
20:    assert Set.equal?(Set.delete(CustomSet.new([3,2,1]), 4), CustomSet.new([1,2,3]))
21:    assert Set.equal?(Set.delete(CustomSet.new([3,2,1]), 2.0), CustomSet.new([1,2,3]))
26:    assert Set.equal?(
27:      Set.difference(CustomSet.new([1,2,3]), CustomSet.new([2,4])),
28:      CustomSet.new([1,3]))
29:    assert Set.equal?(
30:      Set.difference(CustomSet.new([1,2.0,3]), CustomSet.new([2,4])),
31:      CustomSet.new([1,2.0,3]))
36:    assert Set.disjoint?(CustomSet.new([1,2]), CustomSet.new([3,4]))
37:    refute Set.disjoint?(CustomSet.new([1,2]), CustomSet.new([2,3]))
38:    assert Set.disjoint?(CustomSet.new([1.0,2.0]), CustomSet.new([2,3]))
39:    assert Set.disjoint?(CustomSet.new, CustomSet.new)
44:    assert Set.equal?(Set.empty(CustomSet.new([1,2])), CustomSet.new)
45:    assert Set.equal?(Set.empty(CustomSet.new), CustomSet.new)
50:    assert Set.equal?(
51:      Set.intersection(CustomSet.new([:a, :b, :c]), CustomSet.new([:a, :c, :d])),
52:      CustomSet.new([:a, :c]))
53:    assert Set.equal?(
54:      Set.intersection(CustomSet.new([1, 2, 3]), CustomSet.new([1.0, 2.0, 3])),
55:      CustomSet.new([3]))
60:    assert Set.member?(CustomSet.new([1,2,3]), 2)
61:    assert Set.member?(CustomSet.new(1..3), 2)
62:    refute Set.member?(CustomSet.new(1..3), 2.0)
63:    refute Set.member?(CustomSet.new(1..3), 4)
68:    assert Set.equal?(
69:      Set.put(CustomSet.new([1,2,4]), 3),
70:      CustomSet.new([1,2,3,4]))
71:    assert Set.equal?(
72:      Set.put(CustomSet.new([1,2,3]), 3),
73:      CustomSet.new([1,2,3]))
74:    assert Set.equal?(
75:      Set.put(CustomSet.new([1,2,3]), 3.0),
76:      CustomSet.new([1,2,3,3.0]))
81:    assert Set.size(CustomSet.new) == 0
82:    assert Set.size(CustomSet.new([1,2,3])) == 3
83:    assert Set.size(CustomSet.new([1,2,3,2])) == 3
88:    assert Set.subset?(CustomSet.new([1,2,3]), CustomSet.new([1,2,3]))
89:    assert Set.subset?(CustomSet.new([1,2,3]), CustomSet.new([4,1,2,3]))
90:    refute Set.subset?(CustomSet.new([1,2,3]), CustomSet.new([4,1,3]))
91:    refute Set.subset?(CustomSet.new([1,2,3.0]), CustomSet.new([1,2,3,4]))
92:    assert Set.subset?(CustomSet.new, CustomSet.new([4,1,3]))
93:    assert Set.subset?(CustomSet.new, CustomSet.new)
98:    assert Enum.sort(Set.to_list(CustomSet.new)) == []
99:    assert Enum.sort(Set.to_list(CustomSet.new([3,1,2]))) == [1,2,3]
100:    assert Enum.sort(Set.to_list(CustomSet.new([3,1,2,1]))) == [1,2,3]
105:    assert Set.equal?(
106:      CustomSet.union(CustomSet.new([1,3]), CustomSet.new([2])),
107:      CustomSet.new([3,2,1]))
108:    assert Set.equal?(
109:      CustomSet.union(CustomSet.new([1,3]), CustomSet.new([2,3.0])),
110:      CustomSet.new([3.0,3,2,1]))
111:    assert Set.equal?(
112:      CustomSet.union(CustomSet.new([1,3]), CustomSet.new),
113:      CustomSet.new([3,1]))
114:    assert Set.equal?(
115:      CustomSet.union(CustomSet.new, CustomSet.new([2])),
116:      CustomSet.new([2]))
117:    assert Set.equal?(
118:      CustomSet.union(CustomSet.new, CustomSet.new),
119:      CustomSet.new([]))
124:    assert inspect(CustomSet.new) == "%CustomSet{list: []}"
125:    assert inspect(CustomSet.new([1,3,2])) == "%CustomSet{list: [1, 2, 3]}"
128:    assert inspect(CustomSet.new(["A",:b,?C])) == "%CustomSet{list: [67, :b, \"A\"]}"
```

### LINKED ISSUE #180: Changes to Custom Set tests
We recently rewrote the test suite for Custom Set. Since this track implements Custom Set, please [take a look at the new custom_set.json file](https://github.com/exercism/x-common/blob/master/custom-set.json) and see if your track should update its implementation or tests.

The new test suite reorders tests so that students can get to green quickly. It also reduces the number of tests so that students can focus on solving the interesting edge cases.

More details on this change are available in [the pull request](https://github.com/exercism/x-common/pull/232)

### LINKED ISSUE #189: Updated tests for the Custom Set problem
In order to reduce the amount of code required to pass incremental tests (assuming that users pass tests starting from the top), the order of the tests was modified slightly.

Since this track implements Custom Set, please take a look at the new custom-set.json file and see if your track should update its tests.

If you do need to update your tests, please refer to this issue in your PR. That helps us see which tracks still need to update their tests.

If your track is already up to date, go ahead and close this issue.

More details on this change are available in https://github.com/exercism/x-common/pull/257.

### LINKED ISSUE #198: [custom-set] example and test suite diverge
The tests do check for `Inspect` as foolows:

```ex
  test "inspect" do
    assert inspect(CustomSet.new) == "%CustomSet{list: []}"
    assert inspect(CustomSet.new([1,3,2])) == "%CustomSet{list: [1, 2, 3]}"
    assert inspect(CustomSet.new(["A",:b,?C])) == "%CustomSet{list: [67, :b, \"A\"]}"
  end
```

While the example implementation does implement `Inspect` as follows:

```ex
  def inspect(s, opts) do
    concat ["#<CustomSet ", to_doc(CustomSet.to_list(s), opts), ">"]
  end
```

Before commit 9222be9496ae8a9f266edf9c0d91785e77aa3b67 the example and the test matched.

### LINKED ISSUE #217: The custom-set exercism is based around the deprecated Set module and removed Set behaviour
Hello!

The `Set` module was deprecated in 1.2, and the behaviour has since been removed. It's not possible to directly port the changes to the `MapSet` replacement as it's a single data structure implementation rather than a behaviour with a default implementation.

This exercise needs to be removed, or entirely re-written. :(

I notice there are other open issues for this exercise. #198 #180 #121

### PR DESCRIPTION
**Pull request by devonestes**
We have a lot of issues open around the `CustomSet` example. This PR rewrites
that to remove the deprecated `Set` module, and also updates it to use the
current tests listed in the `exercism/x-common` repo.

resolves #121 
resolves #180 
resolves #189 
resolves #198 
resolves #217

### CODE REVIEW

**Comment by parkerl** [comment]
@devonestes excellent work as usual! I think I have a couple small comments. This diff is a little challenging to review. I'm mostly looking at it as a new exercise.

File: exercises/custom-set/custom_set_test.exs [hunk 1]
```diff
@@ -8,121 +8,275 @@ ExUnit.configure exclude: :pending, trace: true
...
-    assert Set.equal?(
-      Set.difference(CustomSet.new([1,2.0,3]), CustomSet.new([2,4])),
-      CustomSet.new([1,2.0,3]))
+    test "sets with elements are not empty" do
+      custom_set = CustomSet.new([1])
+      assert CustomSet.empty?(custom_set) == false
+    end
   end
 
   @tag :pending
```
Comment (parkerl) [inline exercises/custom-set/custom_set_test.exs]: I like the `describe` blocks. Are you at all worried that `@pending` these blocks will make too many test go red at one time? It seems ok since each is focused on one function.

File: exercises/custom-set/example.exs [hunk 2]
```diff
@@ -1,136 +1,74 @@
...
-
-  def empty(_) do
-    new()
-  end
-
-  def equal?(a, b) do
-    unwrap(a) === unwrap(b)
+  @spec new(Enum.t) :: t
+  def new(enumerable) do
+    %CustomSet{map: Map.new(enumerable, fn (x) -> {x, true} end)}
```
Comment (parkerl) [inline exercises/custom-set/example.exs]: I didn't see a test where this behavior of `new` is demonstrated. Perhaps we need a set of tests showing that `[1,1,2,3]` creates a CustomSet of `[1,2,3]`?

File: exercises/custom-set/custom_set_test.exs (hunk 1 above)
Comment (devonestes) [inline exercises/custom-set/custom_set_test.exs]: Frankly, I didn't really think about it too much since that was a refactor. But now that you mention it, I don't think it's a problem. When I was doing the implementation for these, the examples in the tests pretty much all went green at the same time since there aren't really many edge cases to handle in these operations.

File: exercises/custom-set/example.exs (hunk 2 above)
Comment (devonestes) [inline exercises/custom-set/example.exs]: Good call! That wasn't in the `.json` file in the `exercism/x-common` repo so I didn't add it, but I agree that it's a crucial point to test. Should probably be the first one, actually!

I'll add it now.

**Comment by parkerl** [comment]
Looks great! 🎉
